suricata/update/parsers.py

```diff
@@ -0,0 +1,245 @@
+    global_parser = argparse.ArgumentParser(add_help=False)
+
+    for arg, opts in global_arg:
+        global_parser.add_argument(*arg, **opts)
+
+    return global_parser
+
+
+def parse_update(subparsers, global_parser):
+    # The "update" (default) sub-command parser.
+    update_parser = subparsers.add_parser(
+        "update", add_help=True, parents=[global_parser],
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+
+    for arg, opts in update_arg:
+        update_parser.add_argument(*arg, **opts)
+
+    return update_parser
+
+
+def parse_commands(subparsers, global_parser):
+    commands.listsources.register(subparsers.add_parser(
+        "list-sources", parents=[global_parser]))
+    commands.listenabledsources.register(subparsers.add_parser(
+        "list-enabled-sources", parents=[global_parser]))
+    commands.addsource.register(subparsers.add_parser(
+        "add-source", parents=[global_parser]))
+    commands.updatesources.register(subparsers.add_parser(
+        "update-sources", parents=[global_parser]))
+    commands.enablesource.register(subparsers.add_parser(
+        "enable-source", parents=[global_parser]))
+    commands.disablesource.register(subparsers.add_parser(
+        "disable-source", parents=[global_parser]))
+    commands.removesource.register(subparsers.add_parser(
+        "remove-source", parents=[global_parser]))
+
+
+def parse_arg():
+    global_parser = parse_global()
+    global_args, rem = global_parser.parse_known_args()
+
+    if global_args.version:
+        revision_string = " (rev: %s)" % (revision) if revision else ""
+        print("suricata-update version {}{}".format(version, revision_string))
+        sys.exit(0)
+
+    if not rem or rem[0].startswith("-"):
+        rem.insert(0, "update")
+
+    parser = argparse.ArgumentParser()
+    subparsers = parser.add_subparsers(dest="subcommand", metavar="<command>")
+    update_parser = parse_update(subparsers, global_parser)
+
+    update_parser.epilog = r"""other commands:
+    update-sources             Update the source index
+    list-sources               List available sources
+    enable-source              Enable a source from the index
+    disable-source             Disable an enabled source
+    remove-source              Remove an enabled or disabled source
+    list-enabled-sources       List all enabled sources
+    add-source                 Add a new source by URL
+"""
+
+    parse_commands(subparsers, global_parser)
+
+    args = parser.parse_args(rem)
+
+    # Merge global args into args.
+    for arg in vars(global_args):
+        if not hasattr(args, arg):
+            setattr(args, arg, getattr(global_args, arg))
+        elif hasattr(args, arg) and getattr(args, arg) is None:
+            setattr(args, arg, getattr(global_args, arg))
+
+    # Go verbose or quiet sooner than later.
+    if args.verbose:
+        logger.setLevel(logging.DEBUG)
+    if args.quiet:
+        logger.setLevel(logging.WARNING)
```

Comment:
Hmm I'm not sure how is this working but from what I understand, `logger.setLevel` was for the main file's logger instance and since the level was not set there, this should not have worked. Can you figure out why did this work?

Reply:
After reading about the loggers I could understand the following (Please correct me if I am wrong):
1. When loggers are instantiated using `logging.getLogger()` without specifying a name (as `logging.getLogger(name)`)  it applies the configuration to the __root logger__ (in main module) and also to the parser.py module logger due to which `logger.setLevel` (for root logger) here is configured to DEBUG and/or WARNING. So it worked. 
2. If I replace `getLogger()` in the main module by `getLogger(__name__)`, the logger in the parsers.py module will become a child of the root logger and it will thus inherit the root logger configuration and as the root logger has `setLevel` configured to `logging.INFO` it will log INFO messages.
I referred these links:
- https://docs.python.org/3/howto/logging.html#loggers
- https://docs.python.org/3/library/logging.html#logger-objects
- https://stackoverflow.com/questions/50714316/how-to-use-logging-getlogger-name-in-multiple-modules/50715155#50715155

Reply:
Thanks for the explanation. I checked and it seems they are indeed the same root logger instances in both the files. However, since we're doing most of the logger related work in main, I'm a little inclined to setting these levels in the main itself. They are not anyway a part of parser code.